certificates/src/lib.rs

```diff
@@ -108,6 +142,33 @@ impl CertificateValidator {
         // to use with verifying the private key.
         let cert = certs.swap_remove(0);
 
+        if let Some(hostname) = hostname {
```

Comment:
This logic seems reasonable based on how I expect this would work.  Is there some standard that specifies how it's supposed to work?  I'm a little surprised there's no library to do this.

Reply:
Me too! I looked and didn't find one, but maybe I wasn't looking for the right thing. OpenSSL has [`X509_check_host`](https://www.openssl.org/docs/manmaster/man3/X509_check_host.html), which is (I think?) what we want, but AFAICT it isn't exposed by the rust openssl bindings, and I can't find an equivalent in `rustls` or `x509-parser`.

Reply:
seems like `rustls` may use `webpki` for this https://github.com/rustls/rustls/blob/c21eca793b90c74b8c6710416a66a56266abae6f/rustls/src/webpki/verify.rs#L102-L120

Reply:
I looked at `webpki` a bit, and I think it would have similar downsides: we have to re-parse the cert, and the other checks we want to do aren't really supported by `webpki` because we don't have a root cert to pass it for validation.

I realized we can also call `X509_check_host` directly; I'm going to close this PR and rework it to only use OpenSSL and will reopen.